xds/src/main/java/io/grpc/xds/GrpcAuthorizationEngine.java

```diff
@@ -0,0 +1,381 @@
+      this.policies = Collections.unmodifiableList(policies);
+      this.action = action;
+    }
+  }
+
+  /** Instantiated with envoy policyMatcher configuration. */
+  public GrpcAuthorizationEngine(AuthConfig authConfig) {
+    this.authConfig = authConfig;
+  }
+
+  /** Return the auth decision for the request argument against the policies. */
+  public AuthDecision evaluate(Metadata metadata, ServerCall<?,?> serverCall) {
+    checkNotNull(metadata, "metadata");
+    checkNotNull(serverCall, "serverCall");
+    String firstMatch = null;
+    EvaluateArgs args = new EvaluateArgs(metadata, serverCall);
+    for (PolicyMatcher policyMatcher : authConfig.policies) {
+      if (policyMatcher.matches(args)) {
+        firstMatch = policyMatcher.name;
+        break;
+      }
+    }
+    Action decisionType = Action.DENY;
+    if (Action.DENY.equals(authConfig.action) == (firstMatch == null)) {
+      decisionType = Action.ALLOW;
+    }
+    log.log(Level.FINER, String.format("RBAC decision: {%s}, policy match: {%s}",
+        decisionType, firstMatch));
+    return AuthDecision.create(decisionType, firstMatch);
+  }
+
+  /**
+   * Implements a top level {@link Matcher} for a single RBAC policy configuration per envoy
+   * protocol:
+   * https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#config-rbac-v3-policy.
+   *
+   * <p>Currently we only support matching some of the request fields. Those unsupported fields are
+   * considered not match until we stop ignoring them.
+   */
+  static final class PolicyMatcher extends Matcher {
+    private final OrMatcher permissions;
+    private final OrMatcher principals;
+    private final String name;
+
+    /** Constructs a matcher for one RBAC policy. */
+    PolicyMatcher(String name, OrMatcher permissions, OrMatcher principals) {
+      this.name = name;
+      this.permissions = permissions;
+      this.principals = principals;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      return permissions.matches(args) && principals.matches(args);
+    }
+  }
+
+  static final class AuthenticatedMatcher extends Matcher {
+    private final StringMatcher delegate;
+
+    AuthenticatedMatcher(StringMatcher delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    boolean matches(EvaluateArgs args) {
+      if (delegate == null) {
+        return true;
+      }
+      Collection<String> principalNames = args.getPrincipalNames();
+      log.log(Level.FINER, "Matching principal names: " + principalNames);
+      if (principalNames != null) {
+        for (String name : principalNames) {
+          if (delegate.matches(name)) {
+            return true;
+          }
+        }
+      }
+      return false;
+    }
+  }
+
+  static final class DestinationIpMatcher extends Matcher {
+    private final CidrMatcher delegate;
+
+    DestinationIpMatcher(CidrMatcher delegate) {
+      this.delegate = delegate;
```

Comment:
`checkNotNull()` here and in similar places? It will make any nullness bug _much_ easier to track down.